src/lib/raft/raft.c

```diff
@@ -539,14 +592,23 @@ raft_process_heartbeat(struct raft *raft, uint32_t source)
 	 */
 	if (raft->is_write_in_progress)
 		return;
+	raft->leader_last_seen = raft_ev_now(raft_loop());
```

Comment:
Otherwise `leader_idle` will jump `0` in `box.info.election`, while the node might still have `raft_is_leader_seen` == `false`. This is inconsistent. But it's not too important, I think. So let me know if you still want to change that.

Reply:
But why would it have `raft_is_leader_seen` false if you set `leader_last_seen` and `raft_notify_leader_seen` without yields right after each other?

My motivation here is that a leader could be set together with a term bump. While the term is being written, the node will not update last seen timestamp. If a WAL write was long, it could be noticeable.

Reply:
> But why would it have raft_is_leader_seen false if you set leader_last_seen and raft_notify_leader_seen without yields right after each other?

There's some misunderstanding, I guess. I meant, that if I move `leader_last_seen` update above the `is_write_in_progress` check, then during a wal write leader_idle might be zero or close to zero, while leader_is_seen might still be false.

I understood that this is impossible, while writing that. We only react to heartbeats when leader is known. This means we either see it (then it's fine to update last_seen) or we don't see it and this heartbeat is the first in a long time. But then there can't be any wal_writes, because wal_write means new term and new term means that there is no leader.

So, I applied the following diff:
```diff
diff --git a/src/lib/raft/raft.c b/src/lib/raft/raft.c
index 6a831a6e4..82a679d03 100644
--- a/src/lib/raft/raft.c
+++ b/src/lib/raft/raft.c
@@ -601,11 +601,12 @@ raft_process_heartbeat(struct raft *raft, uint32_t source)
                return;
        /*
         * The instance currently is busy with writing something on disk. Can't
-        * react to heartbeats.
+        * react to heartbeats. Still, update leader_last_seen field for the
+        * sake of metrics.
         */
+       raft->leader_last_seen = raft_ev_now(raft_loop());
        if (raft->is_write_in_progress)
                return;
-       raft->leader_last_seen = raft_ev_now(raft_loop());
        /*
         * XXX: it may be expensive to reset the timer like that. It may be less
         * expensive to let the timer work, and remember last timestamp when

```

Reply:
I would propose to cover it with a test like this one:
```C
    raft_node_block(&node);
    is(raft_node_send_leader(
        &node,
        2 /* Term. */
        2 /* Source. */
    ), 0, "leader accepted");
    ok(raft_node_check_full_state(
        &node,
        RAFT_STATE_FOLLOWER /* State. */,
        2 /* Leader. */,
        1 /* Term. */,
        0 /* Vote. */,
        2 /* Volatile term. */,
        0 /* Volatile vote. */,
        "{0: 0}" /* Vclock. */
    ), "in progress of WAL write");
    is(node.raft.leader_last_seen == raft_time());
    raft_run_for(1);
    is(node.raft.leader_last_seen + 1 == raft_time());

    raft_node_send_heartbeat(&node, 2);
    is(node.raft.leader_last_seen == raft_time());
```

Reply:
Thanks! Applied a slightly different test:
```diff
diff --git a/test/unit/raft.c b/test/unit/raft.c
index 98c1ef677..c659bd286 100644
--- a/test/unit/raft.c
+++ b/test/unit/raft.c
@@ -1896,26 +1896,53 @@ raft_test_split_vote(void)
 static void
 raft_test_pre_vote(void)
 {
-	raft_start_test(39);
+	raft_start_test(43);
 	struct raft_node node;
 	raft_node_create(&node);
 
 	/* Check leader_idle calculations. */
+	raft_node_block(&node);
 	is(raft_node_send_leader(
 		&node,
 		2 /* Term. */,
 		2 /* Source. */
 	), 0, "leader notification");
-
+	ok(raft_node_check_full_state(
+		&node,
+		RAFT_STATE_FOLLOWER /* State. */,
+		2 /* Leader. */,
+		1 /* Term. */,
+		0 /* Vote. */,
+		2 /* Volatile term. */,
+		0 /* Volatile vote. */,
+		"{0: 0}" /* Vclock. */
+	), "WAL write is in progress");
 	is(raft_leader_idle(&node.raft), 0, "leader just appeared");
 
 	raft_run_for(node.cfg_death_timeout / 2);
 	is(raft_leader_idle(&node.raft), node.cfg_death_timeout / 2,
 	   "leader_idle increased");
-
 	raft_node_send_heartbeat(&node, 2);
+	is(raft_leader_idle(&node.raft), 0, "heartbeat resets idle counter "
+					    "during WAL write");
 
-	is(raft_leader_idle(&node.raft), 0, "heartbeat resets idle counter");
+	raft_node_unblock(&node);
+	ok(raft_node_check_full_state(
+		&node,
+		RAFT_STATE_FOLLOWER /* State. */,
+		2 /* Leader. */,
+		2 /* Term. */,
+		0 /* Vote. */,
+		2 /* Volatile term. */,
+		0 /* Volatile vote. */,
+		"{0: 1}" /* Vclock. */
+	), "WAL write finished");
+	raft_run_for(node.cfg_death_timeout / 2);
+	is(raft_leader_idle(&node.raft), node.cfg_death_timeout / 2,
+	   "leader_idle increased");
+	raft_node_send_heartbeat(&node, 2);
+	is(raft_leader_idle(&node.raft), 0, "heartbeat resets idle counter "
+					    "when no WAL write");
 
 	raft_node_cfg_is_candidate(&node, false);
 
```